tracing-undertow/src/main/java/com/palantir/tracing/undertow/TracedOperationHandler.java

```diff
@@ -60,11 +61,32 @@ public TracedOperationHandler(HttpHandler delegate, String operation) {
 
     @Override
     public void handleRequest(HttpServerExchange exchange) throws Exception {
-        String traceId = initializeTrace(exchange);
+        HeaderMap headers = exchange.getRequestHeaders();
+        // TODO(rfink): Log/warn if we find multiple headers?
+        String traceIdHeader = headers.getFirst(TRACE_ID); // nullable
+        String traceId;
+        // XXX: again, nasty abstraction leak
+        String requestId;
+
+        // Set up thread-local span that inherits state from HTTP headers
+        if (Strings.isNullOrEmpty(traceIdHeader)) {
```

Comment:
I had to inline the trace creation to get both the span id and the trace id out, without creating a wrapper object. Maybe just reading the MDC here is simpler?

Reply:
Ya, I'd prefer the minimal change where we read the requestId from MDC here.